### PR TITLE
add docs for derive topn from window

### PR DESCRIPTION
**Pull request by windtalker**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.4 (TiDB 6.4 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: derive-topn-from-window.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
+---
+title: 从窗口函数中推导 TopN 或 Limit
+aliases: ['/docs-cn/dev/derive-topn-from-window/']
```
Comment (qiancai) [inline derive-topn-from-window.md:3]:
```suggestion
```

Comment (qiancai) [reply derive-topn-from-window.md:3]: 只有替代某篇已经删除的文档时，才需要加已经删除文档原链接，避免 404

File: derive-topn-from-window.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+---
+
+# 从窗口函数中推导 TopN 或 Limit
+
+窗口函数是一种在 SQL 语句中常见的函数，而对于 ROW_NUMBER() 或者 RANK() 等编号相关的窗口函数，一个常见的用法是在进行窗口函数求值之后，对 ROW_NUMBER() 或者 RANK() 的结果做过滤，例如
```
Comment (qiancai) [inline derive-topn-from-window.md:8]:
```suggestion
[窗口函数](/functions-and-operators/window-functions.md)是一种常见的 SQL 函数。对于 ROW_NUMBER() 或者 RANK() 等编号相关的窗口函数，一种常见的用法是在进行窗口函数求值之后，对求值的结果进行过滤，例如：
```

File: derive-topn-from-window.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+```sql
+SELECT * FROM (SELECT ROW_NUMBER() OVER (ORDER BY a) AS rownumber FROM t) dt WHERE rownumber <= 3
+```
+
+按照正常的 SQL 执行流程，TiDB 需要对 `t` 表的所有数据进行排序后给每一行都求得相应的 `ROW_NUMBER()` 结果之后再进行 `rownumber <= 3` 的过滤，而该优化可以将原始 SQL 等价改写成
```
Comment (qiancai) [inline derive-topn-from-window.md:14]:
```suggestion
按照正常的 SQL 执行流程，TiDB 需要先对 `t` 表的所有数据进行排序，然后为每一行都求得相应的 `ROW_NUMBER()` 结果，最后再进行 `rownumber <= 3` 的过滤。 

从 v7.0.0 开始，TiDB 支持从窗口函数中推导 TopN 或 Limit。通过该功能，TiDB 可以将原始 SQL 等价改写成以下形式：
```

File: derive-topn-from-window.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+```sql
+WITH t_topN AS (SELECT a FROM t1 ORDER BY a LIMIT 3) SELECT * FROM (SELECT ROW_NUMBER() OVER (ORDER BY a) AS rownumber FROM t_topN) dt WHERE rownumber <= 3
+```
+
+可以看出，该优化从窗口函数与后续的过滤条件中推导出了一个 TopN 算子，相比于原始 SQL 中的 Sort 算子，TopN 算子的运行效率远高于 Sort 算子，而且 TiKV/TiFlash 支持 TopN 算子的下推，这能进一步加速改写之后的 SQL 的性能。
```
Comment (qiancai) [inline derive-topn-from-window.md:20]:
```suggestion
可以看出，改写后，TiDB 可以从窗口函数与后续的过滤条件中推导出一个 TopN 算子，相比于原始 SQL 中的 Sort 算子（对应 ORDER BY 函数），TopN 算子的运行效率远高于 Sort 算子，而且 TiKV 和 TiFlash 均支持 TopN 算子的下推，因此这能进一步加速改写之后的 SQL 的性能。
```

File: derive-topn-from-window.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+有两种方法关闭此优化
```
Comment (qiancai) [inline derive-topn-from-window.md:22]:
```suggestion
从窗口函数中推导 TopN 或 Limit 功能默认关闭。如需开启，可以进行以下操作之一：
```

File: derive-topn-from-window.md [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+* 设置 session 变量 [tidb_opt_derive_topn](/system-variables.md#tidb_opt_derive_topn-从-v610-版本开始引入) 为 false
```
Comment (qiancai) [inline derive-topn-from-window.md:24]:
```suggestion
* 设置 session 变量 [tidb_opt_derive_topn](/system-variables.md#tidb_opt_derive_topn-从-v700-版本开始引入) 为 false
```

File: derive-topn-from-window.md [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+* 可参照[优化规则及表达式下推的黑名单](/blocklist-control-plan.md)中的关闭方法。
+
+## 示例
+
+以下通过一些例子对改优化规则进行说明。
```
Comment (qiancai) [inline derive-topn-from-window.md:29]:
```suggestion
以下通过一些例子对该优化规则进行说明。
```

File: derive-topn-from-window.md [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+|       └─Sort_22                    | 3.00     | root      |               | test.t.id1, test.t.value1                                                                                            |
+|         └─TableReader_21           | 3.00     | root      |               | data:TopN_19                                                                                                         |
+|           └─TopN_19                | 3.00     | cop[tikv] |               | partition by test.t.id1 order by test.t.value1, offset:0, count:3                                                    |
+|             └─TableFullScan_18     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                                                       |
++------------------------------------+----------+-----------+---------------+----------------------------------------------------------------------------------------------------------------------+
+```
+
+在该查询中，优化器从窗口函数中推导出来了 TopN 并将它下推给了 TiKV, 值得一提的是这个 TopN 其实是 partition TopN。
```
Comment (qiancai) [inline derive-topn-from-window.md:143]:
```suggestion
在该查询中，优化器从窗口函数中推导出来了 TopN 并将它下推给了 TiKV。需要注意的是，这个 TopN 其实是 partition 的 TopN。
```

File: derive-topn-from-window.md [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+|   └─Shuffle_15                   | 10000.00 | root      |               | execution info: concurrency:5, data sources:[TableReader_13]                                  |
+|     └─Window_11                  | 10000.00 | root      |               | row_number()->Column#7 over(partition by test.t.id1 rows between current row and current row) |
+|       └─Sort_14                  | 10000.00 | root      |               | test.t.id1                                                                                    |
+|         └─TableReader_13         | 10000.00 | root      |               | data:TableFullScan_12                                                                         |
+|           └─TableFullScan_12     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                                |
++----------------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------+
+```
+
+在该查询中，即使 PARTITION 的列是主键的前缀，但是因为主键不是 clustered index，所以 SQL 没被改写。
```
Comment (qiancai) [inline derive-topn-from-window.md:195]:
```suggestion
在该查询中，即使 PARTITION 的列是主键的前缀，但是因为主键不是聚簇索引，所以 SQL 没被改写。
```

File: derive-topn-from-window.md [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+### 限制
+
+* 目前支持改写的窗口函数仅包括 ROW_NUMBER()
```
Comment (qiancai) [inline derive-topn-from-window.md:199]:
```suggestion
* 目前仅有 ROW_NUMBER() 窗口函数支持 SQL 语句改写。
```

File: derive-topn-from-window.md [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+* 当对 ROW_NUMBER() 过滤条件不是 `<` 或者 `<=` 时不支持改写
+* 当过滤条件不是针对 ROW_NUMBER() 时不支持改写
```
Comment (qiancai) [inline derive-topn-from-window.md:201]:
```suggestion
* 只有当 SQL 语句的过滤条件是针对 ROW_NUMBER() 结果而且过滤条件为 `<` 或者 `<=` 时，TiDB 才支持改写 SQL 语句。
```

File: derive-topn-from-window.md [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+可以看出，该优化从窗口函数与后续的过滤条件中推导出了一个 TopN 算子，相比于原始 SQL 中的 Sort 算子，TopN 算子的运行效率远高于 Sort 算子，而且 TiKV/TiFlash 支持 TopN 算子的下推，这能进一步加速改写之后的 SQL 的性能。
+
+有两种方法关闭此优化
+
+* 设置 session 变量 [tidb_opt_derive_topn](/system-variables.md#tidb_opt_derive_topn-从-v700-版本开始引入) 为 `false`
+* 可参照[优化规则及表达式下推的黑名单](/blocklist-control-plan.md)中的关闭方法。
+
+## 示例
+
+以下通过一些例子对该优化规则进行说明。
```
Comment (LittleFall) [inline derive-topn-from-window.md:32]: consider add a plan before this optimazition?

**Review by LittleFall — APPROVED** [review]
others LGTM

File: derive-topn-from-window.md (hunk 12 above)
Comment (windtalker) [reply derive-topn-from-window.md:32]: Already add some plans that can not apply this rule in the following examples.

**Review by joey-yez — APPROVED** [review]

File: derive-topn-from-window.md [hunk 13]
```diff
@@ -0,0 +1,203 @@
+---
+title: 从窗口函数中推导 TopN 或 Limit
```
Comment (qiancai) [inline derive-topn-from-window.md:2]:
```suggestion
title: 从窗口函数中推导 TopN 或 Limit
summary: 介绍从窗口函数中推导 TopN 或 Limit 的优化规则，以及如何开启该规则。
```

File: derive-topn-from-window.md [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+按照正常的 SQL 执行流程，TiDB 需要先对 `t` 表的所有数据进行排序，然后为每一行都求得相应的 `ROW_NUMBER()` 结果，最后再进行 `rownumber <= 3` 的过滤。 
+
+从 v7.0.0 开始，TiDB 支持从窗口函数中推导 TopN 或 Limit。通过该优化规则，TiDB 可以将原始 SQL 等价改写成以下形式：
+
+```sql
+WITH t_topN AS (SELECT a FROM t1 ORDER BY a LIMIT 3) SELECT * FROM (SELECT ROW_NUMBER() OVER (ORDER BY a) AS rownumber FROM t_topN) dt WHERE rownumber <= 3
+```
+
+可以看出，改写后，TiDB 可以从窗口函数与后续的过滤条件中推导出一个 TopN 算子，相比于原始 SQL 中的 Sort 算子（对应 ORDER BY ），TopN 算子的运行效率远高于 Sort 算子，而且 TiKV 和 TiFlash 均支持 TopN 算子的下推，因此这能进一步加速改写之后的 SQL 的性能。
```
Comment (qiancai) [inline derive-topn-from-window.md:21]:
```suggestion
可以看出，改写后，TiDB 可以从窗口函数与后续的过滤条件中推导出一个 TopN 算子，相比于原始 SQL 中的 Sort 算子（对应 `ORDER BY`），TopN 算子的运行效率远高于 Sort 算子，而且 TiKV 和 TiFlash 均支持 TopN 算子的下推，因此这能进一步加速改写之后的 SQL 的性能。
```

File: derive-topn-from-window.md [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+|           └─Limit_23               | 3.00    | cop[tikv] |               | offset:0, count:3                                                                             |
+|             └─TableFullScan_22     | 3.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                                |
++------------------------------------+---------+-----------+---------------+-----------------------------------------------------------------------------------------------+
+```
+
+在该查询中，优化器从窗口函数中推导出来了 Limit 并将它下推给了 TiKV, 值得一提的是这个 Limit 其实是 partition Limit，也就是说对于每个相同 id1 值组成的一组数据上都会进行一次 Limit。
+
+#### 示例 4：包含 ORDER BY 的窗口函数
+
+{{< copyable "sql" >}}
```
Comment (hfxsd) [inline derive-topn-from-window.md:124]:
```suggestion
```

File: derive-topn-from-window.md [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+|           └─TopN_19                | 3.00     | cop[tikv] |               | partition by test.t.id1 order by test.t.value1, offset:0, count:3                                                    |
+|             └─TableFullScan_18     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                                                       |
++------------------------------------+----------+-----------+---------------+----------------------------------------------------------------------------------------------------------------------+
+```
+
+在该查询中，优化器从窗口函数中推导出来了 TopN 并将它下推给了 TiKV。需要注意的是，这个 TopN 其实是 partition 的 TopN。
+
+#### 示例 5：PARTITION BY 列不是主键的前缀
+
+{{< copyable "sql" >}}
```
Comment (hfxsd) [inline derive-topn-from-window.md:151]:
```suggestion
```

File: derive-topn-from-window.md [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+|         └─TableReader_13         | 10000.00 | root      |               | data:TableFullScan_12                                                                            |
+|           └─TableFullScan_12     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                                   |
++----------------------------------+----------+-----------+---------------+--------------------------------------------------------------------------------------------------+
+```
+
+在该查询中，因为 partition 的列不是主键的前缀，所以 SQL 没有被改写。
+
+#### 示例 6：PARTITION BY 列是主键的前缀，但主键不是聚簇索引
+
+{{< copyable "sql" >}}
```
Comment (hfxsd) [inline derive-topn-from-window.md:177]:
```suggestion
```

File: derive-topn-from-window.md [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+|     └─Window_11                  | 10000.00 | root      |               | row_number()->Column#7 over(partition by test.t.id1 rows between current row and current row) |
+|       └─Sort_14                  | 10000.00 | root      |               | test.t.id1                                                                                    |
+|         └─TableReader_13         | 10000.00 | root      |               | data:TableFullScan_12                                                                         |
+|           └─TableFullScan_12     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                                |
++----------------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------+
+```
+
+在该查询中，即使 PARTITION 的列是主键的前缀，但是因为主键不是聚簇索引，所以 SQL 没被改写。
+
+### 限制
```
Comment (hfxsd) [inline derive-topn-from-window.md:201]: 可以把这小节放文档前面部分吗？

Comment (windtalker) [reply derive-topn-from-window.md:189]: 好的

File: derive-topn-from-window.md [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+WITH t_topN AS (SELECT a FROM t1 ORDER BY a LIMIT 3) SELECT * FROM (SELECT ROW_NUMBER() OVER (ORDER BY a) AS rownumber FROM t_topN) dt WHERE rownumber <= 3
+```
+
+可以看出，改写后，TiDB 可以从窗口函数与后续的过滤条件中推导出一个 TopN 算子，相比于原始 SQL 中的 Sort 算子（对应 `ORDER BY`），TopN 算子的运行效率远高于 Sort 算子，而且 TiKV 和 TiFlash 均支持 TopN 算子的下推，因此能进一步提升改写后的 SQL 的性能。
+
+从窗口函数中推导 TopN 或 Limit 默认关闭。你可以通过将 session 变量 [tidb_opt_derive_topn](/system-variables.md#tidb_opt_derive_topn-从-v700-版本开始引入) 设置为 `ON` 开启该功能。
+
+开启后，如需关闭，可以进行以下操作之一：
+
+* 设置 session 变量 [tidb_opt_derive_topn](/system-variables.md#tidb_opt_derive_topn-从-v700-版本开始引入) 为 `false`
```
Comment (hfxsd) [inline derive-topn-from-window.md:28]:
```suggestion
* 设置 session 变量 [tidb_opt_derive_topn](/system-variables.md#tidb_opt_derive_topn-从-v700-版本开始引入) 为 `false`。
```

File: derive-topn-from-window.md [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+* 可参照[优化规则及表达式下推的黑名单](/blocklist-control-plan.md)中的关闭方法。
+
+### 限制
+
+* 目前仅有 ROW_NUMBER() 窗口函数支持 SQL 语句改写。
```
Comment (hfxsd) [inline derive-topn-from-window.md:33]:
```suggestion
* 目前仅 ROW_NUMBER() 窗口函数支持 SQL 语句改写。
```

File: derive-topn-from-window.md [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+|     └─Window_12                    | 3.00    | root      |               | row_number()->Column#6 over(partition by test.t.id1 rows between current row and current row) |
+|       └─Sort_25                    | 3.00    | root      |               | test.t.id1                                                                                    |
+|         └─TableReader_24           | 3.00    | root      |               | data:Limit_23                                                                                 |
+|           └─Limit_23               | 3.00    | cop[tikv] |               | offset:0, count:3                                                                             |
+|             └─TableFullScan_22     | 3.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                                |
++------------------------------------+---------+-----------+---------------+-----------------------------------------------------------------------------------------------+
+```
+
+在该查询中，优化器从窗口函数中推导出来了 Limit 并将它下推给了 TiKV, 值得一提的是这个 Limit 其实是 partition Limit，也就是说对于每个相同 `id1` 值组成的一组数据上都会进行一次 Limit。
```
Comment (hfxsd) [inline derive-topn-from-window.md:119]:
```suggestion
在该查询中，优化器从窗口函数中推导出来了 Limit 并将它下推给了 TiKV。值得一提的是这个 Limit 其实是 partition Limit，也就是说对于每个相同 `id1` 值组成的一组数据上都会进行一次 Limit。
```

**Review by hfxsd — APPROVED** [review]
Rest LGTM

File: derive-topn-from-window.md [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+从窗口函数中推导 TopN 或 Limit 默认关闭。你可以通过将 session 变量 [tidb_opt_derive_topn](/system-variables.md#tidb_opt_derive_topn-从-v700-版本开始引入) 设置为 `ON` 开启该功能。
+
+开启后，如需关闭，可以进行以下操作之一：
+
+* 设置 session 变量 [tidb_opt_derive_topn](/system-variables.md#tidb_opt_derive_topn-从-v700-版本开始引入) 为 `false`。
+* 可参照[优化规则及表达式下推的黑名单](/blocklist-control-plan.md)中的关闭方法。
+
+## 限制
+
+* 目前仅 ROW_NUMBER() 窗口函数支持 SQL 语句改写。
```
Comment (Oreoxmt) [inline derive-topn-from-window.md:33]:
```suggestion
* 目前仅 `ROW_NUMBER()` 窗口函数支持 SQL 语句改写。
```

File: derive-topn-from-window.md [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+* 只有当 SQL 语句的过滤条件是针对 ROW_NUMBER() 结果而且过滤条件为 `<` 或者 `<=` 时，TiDB 才支持改写 SQL 语句。
```
Comment (Oreoxmt) [inline derive-topn-from-window.md:34]:
```suggestion
* 只有当 SQL 语句的过滤条件是针对 `ROW_NUMBER()` 结果而且过滤条件为 `<` 或者 `<=` 时，TiDB 才支持改写 SQL 语句。
```

**Review by fixdb — APPROVED** [review]
+1

File: system-variables.md [hunk 24]
```diff
@@ -2454,6 +2454,14 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
 - 默认值：`3.0`
 - 表示 TiDB 处理一行数据的 CPU 开销。该变量是[代价模型](/cost-model.md)内部使用的变量，不建议修改该变量的值。
 
+### `tidb_opt_derive_topn` <span class="version-mark">从 v7.0.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 类型：布尔型
+- 默认值：`OFF`
+- 表示是否开启[从窗口函数中推导 TopN 或 Limit](/derive-topn-from-window.md) 的优化。
```
Comment (qiancai) [inline system-variables.md:2463]:
```suggestion
- 表示是否开启[从窗口函数中推导 TopN 或 Limit](/derive-topn-from-window.md) 的优化规则。
```

File: derive-topn-from-window.md [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+## 限制
+
+* 目前仅 `ROW_NUMBER()` 窗口函数支持 SQL 语句改写。
+* 只有当 SQL 语句的过滤条件是针对 `ROW_NUMBER()` 结果而且过滤条件为 `<` 或者 `<=` 时，TiDB 才支持改写 SQL 语句。
+
+## 示例
+
+以下通过一些例子对该优化规则进行说明。
+
+### 不带 PARTITION BY 的窗口函数
```
Comment (qiancai) [inline derive-topn-from-window.md:40]:
```suggestion
### 不包含 PARTITION BY 的窗口函数
```

**Review by qiancai — APPROVED** [review]

**Pull request merged**